Description: Add "keepgoing" flag (-k) to not abort on errors
 The rationale for this is it allows for more resilient operation when run from
 an init script.
Forwarded: yes
Author: Rowan Thorpe <rowan@rowanthorpe.com>
Last-Update: 2014-03-03
--- a/src/sflowtool.c
+++ b/src/sflowtool.c
@@ -205,6 +205,9 @@
   int listen4;
   int listen6;
   int listenControlled;
+
+  /* general options */
+  int keepGoing;
 } SFConfig;
 
 /* make the options structure global to the program */
@@ -4238,6 +4241,9 @@
   fprintf(ERROUT, "   -6                 -  listen on IPv6 socket only\n");
   fprintf(ERROUT, "   +4                 -  listen on both IPv4 and IPv6 sockets\n");
   fprintf(ERROUT, "\n");
+  fprintf(ERROUT,"General options:\n");
+  fprintf(ERROUT, "   -k                 -  keep going on non-signal errors rather than aborting\n");
+  fprintf(ERROUT, "\n");
   fprintf(ERROUT, "=============== Advanced Tools ==============================================\n");
   fprintf(ERROUT, "| sFlowTrend (FREE)     - http://www.inmon.com/products/sFlowTrend.php      |\n");
   fprintf(ERROUT, "| Traffic Sentinel      - http://www.inmon.com/support/sentinel_release.php |\n");
@@ -4266,6 +4272,7 @@
   sfConfig.listen4 = NO;
   sfConfig.listen6 = YES;
 #endif
+  sfConfig.keepGoing = NO;
 
   /* walk though the args */
   while (arg < argc) {
@@ -4286,6 +4293,7 @@
 #endif
     case '4':
     case '6':
+    case 'k':
     case '?':
     case 'h': break;
     case 'p':
@@ -4365,6 +4373,7 @@
       sfConfig.listen4 = NO;
       sfConfig.listen6 = YES;
       break;
+    case 'k': sfConfig.keepGoing = YES; break;
     /* remaining are -h or -? */
     default: instructions(*argv); exit(0);
     }
@@ -4453,11 +4462,12 @@
 		    &timeout);
       /* we may return prematurely if a signal was caught, in which case
        * nfds will be -1 and errno will be set to EINTR.  If we get any other
-       * error, abort.
+       * error, abort (unless keepGoing is set).
        */
       if(nfds < 0 && errno != EINTR) {
 	fprintf(ERROUT, "select() returned %d\n", nfds);
-	exit(-9);
+        if(sfConfig.keepGoing == NO)
+          exit(-9);
       }
       if(nfds > 0) {
 	if(soc4 != -1 && FD_ISSET(soc4, &readfds)) readPacket(soc4);
